application/templates/base.html

```diff
@@ -50,47 +59,75 @@
           ga('require', 'eventTracker', {
             attributePrefix: 'data-'
           });
-          {% block google_analytics %}{% endblock %}
+          {% block googleAnalytics %}{% endblock %}
           ga('send', 'pageview');
         </script>
         <script src="/static/vendor/autotrack/autotrack.js"></script>
     {% endif %}
 
-    {% block endhead %} {% endblock %}
-
+    {% block headEnd %}
+      {% if not static_mode %}
+        <link rel="stylesheet" href="/static/stylesheets/{{ 'cms.css' | version_filter }}">
+      {% endif %}
+    {% endblock %}
   </head>
-  <body class="{% block body_classes %}{% endblock %}">
-    <script>document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');</script>
+  
+  <body class="{% block bodyClasses %}{% endblock %}">
+    {% block bodyStart %}
+      <script>document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');</script>
+    {% endblock %}
 
+    {% block skipLink %}
     <div id="skiplink-container">
       <div>
         <a href="#content" class="skiplink">{{ skip_link_message|default('Skip to main content') }}</a>
       </div>
     </div>
+    {% endblock %}
 
-    {% if not static_mode %}
-      {% include 'static_site/_preview.html' %}
-    {% endif %}
-
+    {% block cookieMessage %}
     <div id="global-cookie-message">
       <div class="cookie-message full-width">
         <p>GOV.UK uses cookies to make the site simpler. <a href="/cookies">Find out more about cookies</a></p>
       </div>
     </div>
+    {% endblock %}
 
-    {% include '_shared/_header.html' %}
-
-    {% block after_header %}{% endblock %}
+    {% block header %}
+      {% if not static_mode %}
+        {% include 'static_site/_preview.html' %}
+      {% endif %}
+      
+      {% include '_shared/_header.html' %}
+    {% endblock %}
 
-    <div id="global-header-bar"></div>
+    <div id="global-header-bar"></div> {# This is a hook into CSS that adds a blue bar beneath the top black navbar #}
 
     {% include '_shared/_phase_banner.html' %}
 
-    {% block main_content %}{% endblock %}
+    {% block main %}
+      {% block mainElementDefinition %}<main id="content" class="main">{% endblock %}
+        
+      {% block beforeContent %}
+        {% block breadcrumbs %}
+          {{ render_breadcrumbs(breadcrumbs) }}
+        {% endblock breadcrumbs %}
+        
+        {% block flashMessages %}
+          {{ render_flash_messages(get_flashed_messages(with_categories=true)) }}
+        {% endblock %}
+      {% endblock %}
+      
+      {% block content %}{% endblock %}
+      
+      </main>
+    {% endblock %}
 
-    {% include '_shared/_footer.html' %}
+    {% block footer %}
+      {% include '_shared/_footer.html' %}
+    {% endblock %}
 
-    {% block body_end %}
+    {% block bodyEnd %}
         <script type="text/javascript" src="/static/javascripts/{{ 'all.js' | version_filter }}"></script>
```

Comment:
Should we move this outside of the `bodyEnd` block, so that templates don't have to remember to include `super()`? There’s no reason I can think of why it shouldn't be present on every page.

Reply:
I think it gives a more clear definition of what `bodyEnd` is - it's where all scripts get included in the template.

It allows the developer to choose whether or not they want to include the parents' JS or not.

Reply:
@samuelhwilliams we have no need for anyone to make that choice at the moment though, as all pages need the `all.js` (eg to include polyfills for older browsers). Forgetting to call `super()` feels like it'd be an easy mistake to make, plus it's an extra line of code duplicated in every template...?

Reply:
You're right that it's probably an easy mistake to make. One of my goals/principles with this refactor was to minimise the amount of code that fell outside of clearly-defined blocks, as it makes that impossible to override/modify in derived templates. One thing in particular that is interesting to bear in mind (though I have no _practical_ examples right now), is that we might need to have some JavaScript included specifically before or after the rest of it. The current block structure allows this - whereas if we have the scripts included outside of any block, that isn't really possible/easy.

I think it's clearer, as it is now, to understand that each block comes with some default content. If you want to extend that content, you need to explicitly decide where you want the original content (before/middle/after). If you want to override it, then you discard the content by not including it via `super`. It feels like, as a developer, we need to have a good understanding of how Jinja2 templating works - and this is simply a feature of how the templating language works, which we don't necessarily need to shy away from.

P.S. I'm not _totally_ convinced by my argument. But these are my reasons and I don't think they're totally spurious. Maybe. If you post a reply with reasonable retorts, I'll probably fold and make the changes you want. 😂

Reply:
@samuelhwilliams ha. I'm not sure I have any major retorts, other than that I still feel that `all.js` should always be included in _all_ pages, and always before any other script tags (given that it includes a bunch of polyfills).

But equally, it's not that bad to have to remember to include `super()`, so I'm not too bothered either way. We can always change our mind later... ;-)